stock_reorder_forecast/models/product_product.py

```diff
@@ -0,0 +1,222 @@
+    turnover_period = fields.Integer(
+        string='Turnover period',
+        help='Turnover days to calculate average '
+        'turnover per day. Used by the purchase proposal.')
+    turnover_average = fields.Float(
+        'Turnover per day', digits=get_precision('Purchase Price'),
+        readonly=True, help='Average turnover per day. Used by the purchase '
+        'proposal.')
+    ultimate_purchase = fields.Date(
+        'Ultimate purchase', readonly=True, help='Ultimate date to purchase '
+        'for not running out of stock. Used by the purchase proposal.')
+    automatic_purchase_days = fields.Integer(
+        "Trigger automatic purchase proposal on first supplier",
+        default=0,
+        help="Trigger an automatic RFQ on first (default supplier) if ultimate"
+             "purchase is less than these days away"
+    )
+    is_buy = fields.Boolean("Does the product have Buy Route?",
+                            compute='_compute_isbuy', stored=True)
+
+    @api.multi
+    @api.depends('product_tmpl_id.route_ids')
+    def _compute_isbuy(self):
+        """
+        Returns true if 'Buy' is amongst routes for this product
+        """
+        for this in self:
+            prd_routes = this.product_tmpl_id.route_ids.ids
+            buy_routes = this.product_tmpl_id._get_buy_route()
+            this.is_buy = any(x in buy_routes for x in prd_routes)
+
+    def has_purchase_draft(self):
+        sql = """
+            select count(*) from purchase_order_line
+                where product_id = %s
+                and order_id in (select id from purchase_order where
+                state='draft')
+            """
+        self.env.cr.execute(sql, (self.id, ))
+        return bool(self.env.cr.fetchone()[0])
+
+    def _get_ultimate_purchase(
+            self, stock_period_min, turnover_average):
+        for this in self:
+            stock_days = int(float_round((((
+                this.virtual_available or 0
+                ) - stock_period_min) / turnover_average) + .5, 0))
+            if self.has_purchase_draft():
+                return False
+            if stock_days < 0:
+                return fields.Date.to_string(
+                    date.today())
+            return fields.Date.to_string(
+                date.today() + timedelta(days=stock_days))
+
+    def _get_turnover_period(self):
+        if self.turnover_period:
+            return self.turnover_period
+        # TODO IMPLEMENT SELLER RULES IN FUNCTION _select_seller
+        for seller in self.product_tmpl_id.seller_ids:
+            if seller.name.turnover_period:
+                return seller.name.turnover_period
+        if self.product_tmpl_id.categ_id.turnover_period:
+            return self.product_tmpl_id.categ_id.turnover_period
+        return int(self.env['ir.config_parameter'].search(
+            [('key', '=', 'default_turnover_period')])[0].value)
+
+    @api.model
+    def calc_purchase_date(self):
```

Comment:
multicompany adaptation of the query is not implemented. added to readme.
Currently the cron job gives us global stats, that may be useful for some multicompany configurations , but not all.

https://github.com/OCA/stock-logistics-warehouse/pull/222/commits/d9f90d95492c790e4851748eab4946f3b5ddb19a